### ISSUE TITLE
Difference in struct initialization causing values to be copied between 2 variables

### ISSUE THREAD

**Issue by mikesena** [dotnet/roslyn#58771]
### Description

The code provided in the reproduction steps fails with the following exception, and I'm not able to explain why.  I believe it may be a bug with the framework, but it may be a convention I'm not aware of.

My expectation would be that in all cases provided, the result is the same - a 'new' struct is created / allocated.

```
0 == subjectB.ValueB
Expected: 0, Actual: 5
   at Program.<<Main>$>g__FailsWithSingleInit|0_0() in C:\repos\Program.cs:line 24
   at Program.<Main>$(String[] args) in C:\repos\Program.cs:line 9
```

### Reproduction Steps

The following code will fail on the final assertation in `FailsWithSingleInit`.

I have made a slightly modified version on .NET fiddle, throwing an exception instead.
https://dotnetfiddle.net/LQh6dG

```CSharp
using System.Diagnostics;

PassesNormally();
PassesWithStringAdded();
PassesWhenInitAll();
PassesWhenRunAllViaLocalStaticInit();

// This will fail.
FailsWithSingleInit();



static void FailsWithSingleInit()
{
    // Setup 2 items.
    var subjectA = new ExampleWithStringInitOne();
    var subjectB = new ExampleWithStringInitOne { ValueB = 5, ValueA = 10u };
    Debug.Assert(subjectA.ValueA != subjectB.ValueA, "subjectA.ValueA != subjectB.ValueA", "Expected: {0}, Actual: {1}", subjectA.ValueA, subjectB.ValueA);
    Debug.Assert(subjectA.ValueB != subjectB.ValueB, "subjectA.ValueB != subjectB.ValueB", "Expected: {0}, Actual: {1}", subjectA.ValueB, subjectB.ValueB);

    // Re-assign one.
    subjectB = new ExampleWithStringInitOne { ValueA = 10u };
    Debug.Assert(10u == subjectB.ValueA, "10u == subjectB.ValueA", "Expected: {0}, Actual: {1}", subjectB.ValueA, 10u);
    Debug.Assert(0 == subjectB.ValueB, "0 == subjectB.ValueB", "Expected: {0}, Actual: {1}", 0, subjectB.ValueB);
}

static void PassesNormally()
{
    // Setup 2 items.
    var subjectA = new ExampleWithoutString();
    var subjectB = new ExampleWithoutString { ValueB = 5, ValueA = 10u };
    Debug.Assert(subjectA.ValueA != subjectB.ValueA, "subjectA.ValueA != subjectB.ValueA",
        "Expected: {0}, Actual: {1}", subjectA.ValueA, subjectB.ValueA);
    Debug.Assert(subjectA.ValueB != subjectB.ValueB, "subjectA.ValueB != subjectB.ValueB", "Expected: {0}, Actual: {1}",
        subjectA.ValueB, subjectB.ValueB);

    // Re-assign one.
    subjectB = new ExampleWithoutString { ValueA = 10u };
    Debug.Assert(10u == subjectB.ValueA, "10u == subjectB.ValueA", "Expected: {0}, Actual: {1}", subjectB.ValueA,
        10u);
    Debug.Assert(0 == subjectB.ValueB, "0 == subjectB.ValueB", "Expected: {0}, Actual: {1}", 0, subjectB.ValueB);
}

static void PassesWithStringAdded()
{
    // Setup 2 items.
    var subjectA = new ExampleWithString();
    var subjectB = new ExampleWithString { ValueB = 5, ValueA = 10u };
    Debug.Assert(subjectA.ValueA != subjectB.ValueA, "subjectA.ValueA != subjectB.ValueA", "Expected: {0}, Actual: {1}", subjectA.ValueA, subjectB.ValueA);
    Debug.Assert(subjectA.ValueB != subjectB.ValueB, "subjectA.ValueB != subjectB.ValueB", "Expected: {0}, Actual: {1}", subjectA.ValueB, subjectB.ValueB);

    // Re-assign one.
    subjectB = new ExampleWithString { ValueA = 10u };
    Debug.Assert(10u == subjectB.ValueA, "10u == subjectB.ValueA", "Expected: {0}, Actual: {1}", subjectB.ValueA, 10u);
    Debug.Assert(0 == subjectB.ValueB, "0 == subjectB.ValueB", "Expected: {0}, Actual: {1}", 0, subjectB.ValueB);
}

static void PassesWhenInitAll()
{
    // Setup 2 items.
    var subjectA = new ExampleWithStringInitAll();
    var subjectB = new ExampleWithStringInitAll { ValueB = 5, ValueA = 10u };
    Debug.Assert(subjectA.ValueA != subjectB.ValueA, "subjectA.ValueA != subjectB.ValueA", "Expected: {0}, Actual: {1}", subjectA.ValueA, subjectB.ValueA);
    Debug.Assert(subjectA.ValueB != subjectB.ValueB, "subjectA.ValueB != subjectB.ValueB", "Expected: {0}, Actual: {1}", subjectA.ValueB, subjectB.ValueB);

    // Re-assign one.
    subjectB = new ExampleWithStringInitAll { ValueA = 10u };
    Debug.Assert(10u == subjectB.ValueA, "10u == subjectB.ValueA", "Expected: {0}, Actual: {1}", subjectB.ValueA, 10u);
    Debug.Assert(0 == subjectB.ValueB, "0 == subjectB.ValueB", "Expected: {0}, Actual: {1}", 0, subjectB.ValueB);
}

static void PassesWhenRunAllViaLocalStaticInit()
{
    RunTest<ExampleWithoutString>();
    RunTest<ExampleWithString>();
    RunTest<ExampleWithStringInitOne>();

    static void RunTest<T>() where T : IExample, new()
    {
        
        // Setup 2 items.
        var subjectA = new T();
        var subjectB = new T { ValueB = 5, ValueA = 10u };
        Debug.Assert(subjectA.ValueA != subjectB.ValueA, "subjectA.ValueA != subjectB.ValueA", "Expected: {0}, Actual: {1}", subjectA.ValueA, subjectB.ValueA);
        Debug.Assert(subjectA.ValueB != subjectB.ValueB, "subjectA.ValueB != subjectB.ValueB", "Expected: {0}, Actual: {1}", subjectA.ValueB, subjectB.ValueB);

        // Re-assign one.
        subjectB = new T { ValueA = 10u };
        Debug.Assert(10u == subjectB.ValueA, "10u == subjectB.ValueA", "Expected: {0}, Actual: {1}", subjectB.ValueA, 10u);
        Debug.Assert(0 == subjectB.ValueB, "0 == subjectB.ValueB", "Expected: {0}, Actual: {1}", 0, subjectB.ValueB);
    }
}

internal interface IExample
{
    public uint ValueA { get; init; }
    public int ValueB { get; init; }
}

internal struct ExampleWithoutString : IExample
{
    public uint ValueA { get; init; }
    public int ValueB { get; init; }
}

internal struct ExampleWithString : IExample
{
    public uint ValueA { get; init; }
    public int ValueB { get; init; }
    public string ValueC { get; init; }
}

internal struct ExampleWithStringInitOne : IExample
{
    public uint ValueA { get; init; }
    public int ValueB { get; init; }
    public string ValueC { get; init; } = "x";
}

#pragma warning disable CA1805
internal struct ExampleWithStringInitAll : IExample
{
    public uint ValueA { get; init; } = 0;
    public int ValueB { get; init; } = 0;
    public string ValueC { get; init; } = "x";
}
#pragma warning restore CA1805
```

### Expected behavior

All provided test examples pass, with no assertation issues.

### Actual behavior

The provided code fails with the following:

```
0 == subjectB.ValueB
Expected: 0, Actual: 5
   at Program.<<Main>$>g__FailsWithSingleInit|0_0() in C:\repos\Program.cs:line 24
   at Program.<Main>$(String[] args) in C:\repos\Program.cs:line 9
```

### Regression?

I'm not sure.

### Known Workarounds

The workaround for this case, is to provide a public constructor myself, where all required to be initialised.

### Configuration

.NET 6.0
Windows 10, 21H2, 19044.1415
x64

I don't believe it is specific to this configuration, but I haven't tested it on any other platforms.

### Other information

Reviewing the IL code, it seems to generate a different path for the failing static method.
I believe it's something funky in there.

**Comment by SupinePandora43** [comment]
Is `Regression` term fits here?

**Comment by cston** [comment]
@mikesena, thanks for reporting this issue and for providing the detailed repro.

This is https://github.com/dotnet/roslyn/issues/57870 and has been fixed in 17.1.

With 17.1, the compiler reports errors on the struct declaration that all fields must be assigned:
```
(113,17): error CS0843: Auto-implemented property 'ExampleWithStringInitOne.ValueA'
    must be fully assigned before control is returned to the caller.
(113,17): error CS0843: Auto-implemented property 'ExampleWithStringInitOne.ValueB'
    must be fully assigned before control is returned to the caller.
```
